src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -625,5 +626,112 @@ public override async Task<VSInternalWrapWithTagResponse> RazorWrapWithTagAsync(
 
             return request?.Response;
         }
+
+        public override async Task<RazorFoldingRangeResponse?> ProvideFoldingRangesAsync(RazorFoldingRangeRequestParam foldingRangeParams, CancellationToken cancellationToken)
+        {
+            if (foldingRangeParams is null)
+            {
+                throw new ArgumentNullException(nameof(foldingRangeParams));
+            }
+
+            var csharpRanges = new List<OmniSharp.Extensions.LanguageServer.Protocol.Models.FoldingRange>();
+            var csharpDocument = GetCSharpDocumentSnapshsot(foldingRangeParams.TextDocument.Uri.ToUri());
+            var csharpTask = Task.CompletedTask;
+            if (csharpDocument is not null)
+            {
+                csharpTask = Task.Run(async () =>
+                    {
+                        var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync(
+                        foldingRangeParams.HostDocumentVersion, csharpDocument, cancellationToken);
+
+                        if (synchronized)
+                        {
+                            var csharpRequestParams = new FoldingRangeParams()
+                            {
+                                TextDocument = new()
+                                {
+                                    Uri = csharpDocument.Uri
+                                }
+                            };
+
+                            var request = await _requestInvoker.ReinvokeRequestOnServerAsync<FoldingRangeParams, OmniSharp.Extensions.LanguageServer.Protocol.Models.FoldingRange[]>(
+                                Methods.TextDocumentFoldingRange.Name,
+                                RazorLSPConstants.RazorCSharpLanguageServerName,
+                                SupportsFoldingRange,
+                                csharpRequestParams,
+                                cancellationToken).ConfigureAwait(false);
+
+                            var result = request.Result;
+                            if (result is null)
+                            {
+                                throw new InvalidOperationException("CSharp result is null");
```

Comment:
not sure about this... still figuring out what "retry" would look like here. What is the actual case we're looking for to know that the CSharp server didn't give us information?

Reply:
cc @NTaylorMullen

Reply:
If the request throws we could probably categorize as unexpected / should retry. A try/catch combined with "have we ever gotten C# folding ranges" might enable us to build a good retry algorithm.

Reply:
Alright, so I'll remove the null -> throw logic and just let the request invoker throw here? null result is just nothing?

Reply:
Sounds great to me!